File: 957-minimum-add-to-make-parentheses-valid/minimum-add-to-make-parentheses-valid.java
class Solution {

    // public Boolean closingParantheses(char a , char b){



    // }

    public int minAddToMakeValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        
        if(s.length() == 0){
            return 0;
        }

        for(int i = 0 ; i<s.length() ; i++){

            char ch = s.charAt(i);
            
            if(stack.empty()){
                stack.push(ch);
                continue;
            }

            if(stack.peek() == '(' && ch == ')'){
                stack.pop();
            }
            else{
                stack.push(ch);
            }

        }

        return stack.size();

    }
}

File: 957-minimum-add-to-make-parentheses-valid/minimum-add-to-make-parentheses-valid.py
class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        
        stack = []

        for ch in s:
            if(not stack):
                stack.append(ch)
                continue
            if stack[-1] == '(' and ch == ')':
                stack.pop()
            else:
                stack.append(ch)
        return len(stack)